File: C++/Sorting.cpp
#include <iostream>
#include <cstdlib>
using namespace std;

#define SIZE 25
#define RANGE 10000

int numbers[SIZE];

void init_number(){
    srand(time(0));
    for(int index = 0; index < SIZE; index++){
        numbers[index] = rand() % (RANGE + 1);
    }
}
void swap(int array[], int i, int j){
    int temp = array[i];
    array[i] = array[j];
    array[j] = temp;
}

int get_min(int array[], int i, int j){
    int n = SIZE;
    int min_value = array[i];
    int min_index = i;
    for(int index = i; index < n; index++){
        if (min_value > array[index]){
            min_value = array[index];
            min_index = index;
        }
    }
    return min_index;
}

void print_array(int* array){
    cout << endl <<"[ ";
    for(int index = 0; index < SIZE; index++){
        cout << array[index];
        if (index == SIZE - 1) break;
        cout << ", ";
    }
    cout << " ]" << endl;
}

int* bubble_sort(int array[]){
    int n = SIZE;
    for(int i = 0; i < n; i++){
        for (int j = i + 1; j < n; j++){
            if (array[i] > array[j]){
                swap(array, i, j);
            }
        }
    }
    return array;
}

int* copy_array(int array[]){
    int n = SIZE;
    int* array2 = (int*) malloc(n * sizeof(int));
    for(int i = 0; i < n; i++){
        array2[i]= array[i];
    }
    return array2;
}

int* selection_sort(int array[]){
    int n = SIZE;
    for(int i = 0; i < n; i++){
        int min_index = get_min(array, i, n);
        swap(array, i, min_index);
    }
    return array;
}

int* insertion_sort(int array[]){
    int n = SIZE;
    for(int i = 0; i < n; i++){
        for(int j = i; j > 0; j--){
            if(array[j] < array[j - 1]){
                swap(array, j - 1, j);
            } else{
                break;
            }
        }
    }
    return array;
}

int* split(int array[], int start, int end) {
    int size = end - start + 1;
    int* result = (int*) malloc(sizeof(int) * size);
    for (int i = 0; i < size; i++) {
        result[i] = array[start + i];
    }
    return result;
}

int* merge(int arr1[], int arr2[], int len1, int len2) {
    int* result = (int*) malloc(sizeof(int) * (len1 + len2));
    int i = 0, j = 0, index = 0;
    while (i < len1 && j < len2) {
        if (arr1[i] < arr2[j]) {
            result[index++] = arr1[i++];
        } else {
            result[index++] = arr2[j++];
        }
    }
    while (i < len1) result[index++] = arr1[i++];
    while (j < len2) result[index++] = arr2[j++];
    free(arr1);
    free(arr2);
    return result;
}

int* merge_sort(int array[], int n) {
    if (n < 2) {
        int* result = (int*) malloc(sizeof(int) * n);
        for (int i = 0; i < n; i++) result[i] = array[i];
        return result;
    }
    int len1 = (n + 1) / 2;
    int len2 = n / 2;
    int* arr1 = split(array, 0, len1 - 1);
    int* arr2 = split(array, len1, n - 1);
    return merge(merge_sort(arr1, len1), merge_sort(arr2, len2),len1,len2);
}

int* tree_sort(int array[]){
    class node{
        public:
        int val;
        node* right;
        node* left;
        node(int val_f, node* right_f, node* left_f){
            val = val_f;
            right = right_f;
            left = left_f;
        }
        void insert(int val_f){
            if (val < val_f){
                if (right != NULL) right->insert(val_f);
                else right = new node(val_f, NULL, NULL);
            }else {
                if (left != NULL) left->insert(val_f);
                else left = new node(val_f, NULL, NULL);
            }
        }
        // Return slot filled
        int traverse(int* array, int index){
            int i = index;
            if (left != NULL) i = left->traverse(array, i) + 1;
            array[i] = val;
            // cout << i << endl;
            if (right != NULL) i = right->traverse(array, i + 1);
            return i;
        }
    };
    node* root = new node(array[0], NULL, NULL);
    int n = SIZE;
    for(int i = 1; i < n; i++){
        root->insert(array[i]);
    }
    // cout<<endl<<root->val<<" "<<root->right->val<<" "<<root->left->val;
    root->traverse(array,0);
    return array;
}

int partition(int* array, int start, int end){
    if (start >= end) return start;
    int pivot = start;
    int left = start + 1;
    int right = end;
    while (left <= right){
        if (array[left] > array[pivot] && array[right] < array[left]) swap(array, left, right);
        if (array[left] <= array[pivot]) left += 1;
        if (array[right] >= array[pivot]) right -= 1;
    }
    swap(array,pivot, right);
    return right;
}

int* quicksort(int* array, int start, int end){
    if (start >= end) return array;
    int pivot = partition(array, start, end);
    quicksort(array, start, pivot - 1);
    quicksort(array, pivot + 1, end);
    return array;
}


int main()
{
    init_number();
    cout << "Numbers: ";
    print_array(numbers);
    cout << "Result of BubbleSort:";
    print_array(bubble_sort(copy_array(numbers)));
    cout << "Result of SelectionSort:";
    print_array(selection_sort(copy_array(numbers)));
    cout << "Result of InsertionSort:";
    print_array(insertion_sort(copy_array(numbers)));
    cout << "Result of MergeSort:";
    print_array(merge_sort(copy_array(numbers), SIZE));
    cout << "Result of TreeSort:";
    print_array(tree_sort(copy_array(numbers)));
    cout << "Result of Quicksort:";
    print_array(quicksort(copy_array(numbers), 0, SIZE - 1));
    return 0;
}

File: Python/sorting.py
def bubble_sort(arr):
    for i in range(len(arr)):
        for j in range(i, len(arr)):
            if arr[i] > arr[j]:
                arr[i], arr[j] = arr[j], arr[i]
    return arr


def selection_sort(arr):
    new_arr = []
    while arr:
        min_arr = min(arr)
        new_arr.append(min_arr)
        arr.remove(min_arr)
    return new_arr


def insertion_sort(arr):
    new_arr = []
    for el in arr:
        index = 0
        for i in range(len(new_arr)):
            if el > new_arr[i]:
                index += 1
        new_arr.insert(index, el)
    return new_arr


def merge_sort(arr):
    def merge(arr1, arr2, result=[]):
        if not arr1:
            result.extend(arr2)
            return result
        if not arr2:
            result.extend(arr1)
            return result
        if arr1[0] > arr2[0]:
            result.append(arr2[0])
            return merge(arr1, arr2[1:], result)
        else:
            result.append(arr1[0])
            return merge(arr1[1:], arr2, result)

    def split(arr1):
        length = len(arr1)
        return [arr1[:length // 2], arr1[length // 2:]]

    # return merge(*split(arr))
    if len(arr) == 1:
        return arr
    arr1, arr2 = split(arr)
    return merge(merge_sort(arr1), merge_sort(arr2))


def quick_sort(arr):
    if len(arr) < 2:
        return arr

    def partition(arr1):
        pivot = arr1[0]
        return [x for x in arr1 if x < pivot], pivot, [x for x in arr1 if x > pivot]

    arr1, pivot, arr2 = partition(arr)
    # return arr1, arr2
    return [*quick_sort(arr1), pivot, *quick_sort(arr2)]


def tree_sort(arr):
    class Node:
        def __init__(self, val_f):
            self.val = val_f
            self.right = None
            self.left = None

        def insert(self, val_f):
            if self.val < val_f:
                if self.right:
                    self.right.insert(val_f)
                else:
                    self.right = Node(val_f)
            else:
                if self.left:
                    self.left.insert(val_f)
                else:
                    self.left = Node(val_f)

        def traverse(self, arr):
            if self.left:
                self.left.traverse(arr)
            arr.append(self.val)
            if self.right:
                self.right.traverse(arr)
            # return arr

    root = Node(arr[0])
    new_arr = []
    for num in arr[1:]:
        root.insert(num)
    root.traverse(new_arr)
    return new_arr


def heap_sort(arr):
    def heapify(arr1, limit, index):
        largest, left, right = index, (2 * index) + 1, (2 * index) + 2
        if left < limit and arr1[left] > arr1[largest]:
            largest = left
        if right < limit and arr1[right] > arr1[largest]:
            largest = right
        if largest != index:
            arr1[index], arr1[largest] = arr1[largest], arr1[index]
            heapify(arr1, limit, largest)

    for i in range(len(arr) // 2 - 1, -1, -1):
        heapify(arr, len(arr), i)
    # return arr
    for i in range(len(arr)):
        arr[0], arr[-i - 1] = arr[-i - 1], arr[0]
        heapify(arr, len(arr) - i -1, 0)
    return arr


import random
length = 20
RANGE = 10_000
numbers = [random.randint(1, RANGE) for i in range(length)]
print("Numbers:", numbers)
print("Result of bubble sort: ", bubble_sort(numbers.copy()))
print("Result of selection sort: ", selection_sort(numbers.copy()))
print("Result of insertion sort: ", insertion_sort(numbers.copy()))
print("Result of merge sort: ", merge_sort(numbers.copy()))
print("Result of quick sort: ", quick_sort(numbers.copy()))
print("Result of tree sort: ", tree_sort(numbers.copy()))
print("Result of heap sort: ", heap_sort(numbers.copy()))